File: frontend/src/collectionQuality.ts
import { requestJSON } from "./api.js";
import { readCollection, type Collection } from "./collections.js";
import { parseQuery, type Query } from "./query.js";

export interface CoverageCounts { complete:number; partial:number; failed:number; unprocessed:number; none:number }
export interface ProcessingCoverage {
  configuration:"configured"|"unconfigured"|"profile_required";
  profile:string; profiles:string[]; profile_fingerprint:string; generation_id:string;
  counts:CoverageCounts|null;
}
export interface QualityDimension { field:string; values:{value:string;count:number}[]; missing:number; other:number }
export interface CollectionQuality {
  collection:Collection & {coverage:ProcessingCoverage}; source_fingerprint:string;
  dimensions:QualityDimension[]; zero_bytes:number; mismatches:number;
  duplicate_documents:number; spikes:{field:string;value:string;count:number}[];
}
const fields=new Set(["extension","media_type","media_family","modified_month","size","text_coverage","duplicates"]);
const states=["complete","partial","failed","unprocessed","none"] as const;
const uuid=/^[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/;
const digest=/^[0-9a-f]{64}$/;
function check(value:unknown):asserts value {if(!value)throw new Error("Invalid collection quality receipt. Reload before continuing.");}
function object(value:unknown):Record<string,unknown>{check(value!==null&&typeof value==="object"&&!Array.isArray(value));return value as Record<string,unknown>;}
function count(value:unknown,total:number):value is number{return typeof value==="number"&&Number.isSafeInteger(value)&&value>=0&&value<=total;}

export function readCoverage(value:unknown,total:number):ProcessingCoverage {
  const raw=object(value);
  check(["configured","unconfigured","profile_required"].includes(String(raw.configuration)));
  check(typeof raw.configuration==="string"&&typeof raw.profile==="string"&&raw.profile.length<=128);
  check(Array.isArray(raw.profiles)&&raw.profiles.length<=64&&raw.profiles.every(x=>typeof x==="string"&&x.length>0&&x.length<=128));
  check(new Set(raw.profiles).size===raw.profiles.length);
  check(typeof raw.profile_fingerprint==="string"&&typeof raw.generation_id==="string"&&raw.generation_id.length<=1024);
  if(raw.configuration==="configured"){
    check(raw.profiles.includes(raw.profile)&&digest.test(raw.profile_fingerprint));
    const counts=object(raw.counts);
    check(Object.keys(counts).length===states.length&&states.every(key=>count(counts[key],total)));
    check(states.reduce((sum,key)=>sum+(counts[key] as number),0)===total);
  }else check(raw.counts===null&&raw.profile===""&&raw.profile_fingerprint==="");
  return raw as unknown as ProcessingCoverage;
}

export async function collectionQuality(session:string,id:string,profile:string,signal:AbortSignal):Promise<CollectionQuality>{
  check(uuid.test(id));
  const params=new URLSearchParams();if(profile)params.set("profile",profile);
  const suffix=params.size?`?${params}`:"";
  const raw=object(await requestJSON<unknown>(`/api/v1/collections/${id}/quality${suffix}`,session,{signal}));
  const collection=readCollection(raw.collection);
  check(collection.id===id);
  const coverage=readCoverage(object(raw.collection).coverage,collection.file_count);
  if(profile)check(coverage.profile===profile);
  check(typeof raw.source_fingerprint==="string"&&digest.test(raw.source_fingerprint));
  check(Array.isArray(raw.dimensions)&&raw.dimensions.length<=7);
  const seen=new Set<string>();
  for(const value of raw.dimensions){
    const d=object(value);check(typeof d.field==="string"&&fields.has(d.field)&&!seen.has(d.field));seen.add(d.field);
    check(Array.isArray(d.values)&&d.values.length<=50&&count(d.missing,collection.file_count)&&count(d.other,collection.file_count));
    let total=d.missing+d.other;const keys=new Set<string>();
    for(const bucket of d.values){const b=object(bucket);check(typeof b.value==="string"&&b.value.length<=1024&&!keys.has(b.value)&&count(b.count,collection.file_count));keys.add(b.value);total+=b.count;}
    check(total===collection.file_count);
  }
  for(const key of ["zero_bytes","mismatches","duplicate_documents"])check(count(raw[key],collection.file_count));
  check(Array.isArray(raw.spikes)&&raw.spikes.length<=7);
  for(const spike of raw.spikes){const s=object(spike);check(typeof s.field==="string"&&fields.has(s.field)&&typeof s.value==="string"&&s.value.length<=1024&&count(s.count,collection.file_count));}
  return {...raw,collection:{...collection,coverage}} as unknown as CollectionQuality;
}

export function qualitySuggestion(id:string,field:string,values:string[]):Query {
  check(uuid.test(id)&&["extension","media_type","media_family","text_coverage","duplicates"].includes(field));
  check(values.length>0&&values.length<=32&&values.every(value=>value.length>0&&value.length<=1024));
  if(field==="duplicates")check(values.every(value=>value==="duplicate"||value==="unique"));
  const operand=field==="media_type"?"mime":field==="duplicates"?"has_duplicates":field;
  const terms=values.map(value=>`${operand}:${JSON.stringify(field==="duplicates"?(value==="duplicate"?"true":"false"):value)}`);
  return parseQuery(JSON.stringify({v:1,text:`(${terms.join(" OR ")})`,syntax:"advanced",mode:"lexical",filters:{collection_ids:[id]},sort:{field:"name",direction:"asc"}}));
}

File: internal/store/collection_quality.go
package store

import (
	"context"
	"crypto/sha256"
	"database/sql"
	"encoding/hex"
	"encoding/json/v2"
	"errors"
	"fmt"
	"mime"
	"path"
	"slices"
	"strings"
	"time"

	"go.kenn.io/docbank/internal/query"
)

// QualityBucket is a frequency value in a bounded collection census.
type QualityBucket struct {
	Value string
	Count int64
}

// QualityDimension retains the top values and accounts for every other member.
type QualityDimension struct {
	Field          string
	Values         []QualityBucket
	Missing, Other int64
}

// QualitySpike identifies a descriptive concentration in one dimension.
type QualitySpike struct {
	Field, Value string
	Count        int64
}

// CollectionQuality is an aggregate receipt from one bounded source snapshot.
type CollectionQuality struct {
	Collection                                Collection
	SourceFingerprint                         string
	Dimensions                                []QualityDimension
	ZeroBytes, Mismatches, DuplicateDocuments int64
	Spikes                                    []QualitySpike
}

var (
	// ErrQualityTooLarge rejects a census that cannot be returned in full.
	ErrQualityTooLarge = errors.New("collection quality too large")
	// ErrQualityUnavailable reports an interrupted or timed-out census.
	ErrQualityUnavailable = errors.New("collection quality unavailable")
	// ErrInvalidQualityFields rejects unknown or repeated dimensions.
	ErrInvalidQualityFields = errors.New("invalid collection quality fields")
)

const (
	maxCollectionQualityMembers = 250_000
	maxCollectionQualityBytes   = 64 << 20
	collectionQualityTimeout    = 5 * time.Second
)

var collectionQualityFields = []string{"duplicates", "extension", "media_family", "media_type", "modified_month", "size", "text_coverage"}

func normalizeQualityFields(fields []string) ([]string, error) {
	if len(fields) == 0 {
		return slices.Clone(collectionQualityFields), nil
	}
	if len(fields) > len(collectionQualityFields) {
		return nil, ErrInvalidQualityFields
	}
	result := slices.Clone(fields)
	slices.Sort(result)
	for i, field := range result {
		if !slices.Contains(collectionQualityFields, field) || i > 0 && result[i-1] == field {
			return nil, fmt.Errorf("%w: %q", ErrInvalidQualityFields, field)
		}
	}
	return result, nil
}

// CollectionQuality reads the full census before deriving any successful
// histogram. The timeout and both limits apply even when few fields are requested.
func (s *Store) CollectionQuality(ctx context.Context, id string, selection CoverageSelection, fields []string) (result CollectionQuality, retErr error) {
	selection, err := normalizeCoverageSelection([]CoverageSelection{selection})
	if err != nil {
		return CollectionQuality{}, err
	}
	fields, err = normalizeQualityFields(fields)
	if err != nil {
		return CollectionQuality{}, err
	}
	if err := validateUUIDv4(id); err != nil {
		return CollectionQuality{}, ErrNotFound
	}
	ctx, cancel := context.WithTimeout(ctx, collectionQualityTimeout)
	defer cancel()
	defer func() {
		if ctx.Err() != nil {
			result = CollectionQuality{}
			retErr = errors.Join(ErrQualityUnavailable, ctx.Err(), retErr)
		}
	}()
	tx, err := s.db.BeginTx(ctx, &sql.TxOptions{ReadOnly: true})
	if err != nil {
		return CollectionQuality{}, err
	}
	defer func() { _ = tx.Rollback() }()
	census, err := collectionQualityCensusTx(ctx, tx, id, selection)
	if err != nil {
		return CollectionQuality{}, err
	}
	result, err = aggregateCollectionQuality(ctx, census, fields)
	if err != nil {
		return CollectionQuality{}, err
	}
	if err := tx.Commit(); err != nil {
		return CollectionQuality{}, err
	}
	return result, nil
}

// The census/aggregation boundary allows a cache to compare fresh source
// fingerprints before reusing an immutable receipt. Census rows never escape
// the store API and contain no document text or provider error payloads.
type collectionQualityCensus struct {
	collection  Collection
	fingerprint string
	rows        []collectionQualityRow
}

type collectionQualityRow struct {
	NodeID                                                                int64
	VersionID, BlobHash, Name, MediaType, ModifiedAt                      string
	Size                                                                  int64
	State, AttachmentID, PublishedAt, BuildID, Completeness               string
	PartialSuccess, Truncated, LexicalSegmentCount, Verified, Serving     int64
	WaiterID, JobID, WaiterState, JobState, WaiterUpdatedAt, JobUpdatedAt string
	DuplicateCount                                                        int64
}

const qualityProjectionCTE = `quality_projection AS (
 SELECT p.*,n.name,COALESCE(v.mime_type,'') media_type,n.modified_at,v.size,
  COALESCE(d.references_count,1) duplicate_count
 FROM processing_coverage p JOIN nodes n ON n.id=p.node_id
 JOIN content_versions v ON v.version_id=p.version_id
 LEFT JOIN duplicate_counts d ON d.blob_hash=p.blob_hash
)`

// Projection size includes fixed-width numeric storage and every selected
// string's UTF-8 bytes. Check before scanning strings into application memory.
const qualityProjectionBytes = `128+length(CAST(version_id AS BLOB))+length(CAST(blob_hash AS BLOB))+
 length(CAST(name AS BLOB))+length(CAST(media_type AS BLOB))+length(CAST(modified_at AS BLOB))+
 length(CAST(state AS BLOB))+length(CAST(attachment_id AS BLOB))+length(CAST(published_at AS BLOB))+
 length(CAST(build_id AS BLOB))+length(CAST(completeness AS BLOB))+length(CAST(waiter_id AS BLOB))+
 length(CAST(job_id AS BLOB))+length(CAST(waiter_state AS BLOB))+length(CAST(job_state AS BLOB))+
 length(CAST(waiter_updated_at AS BLOB))+length(CAST(job_updated_at AS BLOB))`

func collectionQualityCensusTx(ctx context.Context, q metadataQuerier, id string, selection CoverageSelection) (collectionQualityCensus, error) {
	collection, err := collectionSummaryByID(ctx, q, id)
	if err != nil {
		return collectionQualityCensus{}, err
	}
	if err := validateCollectionQualityBounds(collection.FileCount, 0); err != nil {
		return collectionQualityCensus{}, err
	}
	generation, err := collectionGenerationTx(ctx, q)
	if err != nil {
		return collectionQualityCensus{}, err
	}
	coverage, err := collectionCoverageTx(ctx, q, []string{id}, selection, generation)
	if err != nil {
		return collectionQualityCensus{}, err
	}
	collection.Coverage = coverage[id]
	cte := `WITH ` + CollectionMembershipCTE + `,` + CurrentContentMembershipCTE + `,
 coverage_members AS (SELECT node_id FROM collection_members WHERE ingest_id=?),
 ` + processingCoverageCTE() + `,
 duplicate_counts AS (SELECT blob_hash,COUNT(*) references_count FROM current_content_members GROUP BY blob_hash),
 ` + qualityProjectionCTE
	args := []any{id, selection.ProfileFingerprint, generation}
	var projectedBytes int64
	if err := q.QueryRowContext(ctx, cte+` SELECT COALESCE(SUM(`+qualityProjectionBytes+`),0) FROM quality_projection`, args...).Scan(&projectedBytes); err != nil {
		return collectionQualityCensus{}, err
	}
	if err := validateCollectionQualityBounds(collection.FileCount, projectedBytes); err != nil {
		return collectionQualityCensus{}, err
	}
	rows, err := q.QueryContext(ctx, cte+` SELECT node_id,version_id,blob_hash,name,media_type,modified_at,size,
 state,attachment_id,published_at,build_id,completeness,partial_success,truncated,lexical_segment_count,verified,serving,
 waiter_id,job_id,waiter_state,job_state,waiter_updated_at,job_updated_at,duplicate_count
 FROM quality_projection ORDER BY node_id`, args...)
	if err != nil {
		return collectionQualityCensus{}, err
	}
	defer func() { _ = rows.Close() }()
	census := collectionQualityCensus{collection: collection, rows: make([]collectionQualityRow, 0, int(collection.FileCount))}
	hash := sha256.New()
	if err := json.MarshalWrite(hash, collection, json.Deterministic(true)); err != nil {
		return collectionQualityCensus{}, err
	}
	for rows.Next() {
		if err := ctx.Err(); err != nil {
			return collectionQualityCensus{}, err
		}
		var row collectionQualityRow
		if err := rows.Scan(&row.NodeID, &row.VersionID, &row.BlobHash, &row.Name, &row.MediaType, &row.ModifiedAt, &row.Size,
			&row.State, &row.AttachmentID, &row.PublishedAt, &row.BuildID, &row.Completeness, &row.PartialSuccess, &row.Truncated, &row.LexicalSegmentCount, &row.Verified, &row.Serving,
			&row.WaiterID, &row.JobID, &row.WaiterState, &row.JobState, &row.WaiterUpdatedAt, &row.JobUpdatedAt, &row.DuplicateCount); err != nil {
			return collectionQualityCensus{}, err
		}
		if err := json.MarshalWrite(hash, row, json.Deterministic(true)); err != nil {
			return collectionQualityCensus{}, err
		}
		census.rows = append(census.rows, row)
	}
	if err := rows.Err(); err != nil {
		return collectionQualityCensus{}, err
	}
	if err := rows.Close(); err != nil {
		return collectionQualityCensus{}, err
	}
	if int64(len(census.rows)) != collection.FileCount {
		return collectionQualityCensus{}, fmt.Errorf("%w: incomplete current-member census", ErrQualityUnavailable)
	}
	census.fingerprint = hex.EncodeToString(hash.Sum(nil))
	return census, nil
}

func validateCollectionQualityBounds(members, bytes int64) error {
	if members > maxCollectionQualityMembers || bytes > maxCollectionQualityBytes {
		return ErrQualityTooLarge
	}
	return nil
}

func aggregateCollectionQuality(ctx context.Context, census collectionQualityCensus, fields []string) (CollectionQuality, error) {
	result := CollectionQuality{Collection: census.collection, SourceFingerprint: census.fingerprint,
		Dimensions: make([]QualityDimension, 0, len(fields)), Spikes: make([]QualitySpike, 0)}
	counts := make([]map[string]int64, len(fields))
	missing := make([]int64, len(fields))
	for i := range counts {
		counts[i] = make(map[string]int64)
	}
	for _, row := range census.rows {
		if err := ctx.Err(); err != nil {
			return CollectionQuality{}, err
		}
		if row.Size == 0 {
			result.ZeroBytes++
		}
		mimeFamily := query.ClassifyMedia(row.MediaType, "")
		extFamily := query.ClassifyMedia("", row.Name)
		if mimeFamily != "unknown" && extFamily != "unknown" && mimeFamily != extFamily {
			result.Mismatches++
		}
		if row.DuplicateCount > 1 {
			result.DuplicateDocuments++
		}
		for i, field := range fields {
			value := qualityValue(row, field, census.collection.Coverage.Configuration == "configured")
			if value == "" {
				missing[i]++
			} else {
				counts[i][value]++
			}
		}
	}
	for i, field := range fields {
		dimension := QualityDimension{Field: field, Values: make([]QualityBucket, 0, len(counts[i])), Missing: missing[i]}
		for value, count := range counts[i] {
			dimension.Values = append(dimension.Values, QualityBucket{value, count})
		}
		slices.SortFunc(dimension.Values, func(a, b QualityBucket) int {
			if a.Count > b.Count {
				return -1
			}
			if a.Count < b.Count {
				return 1
			}
			return strings.Compare(a.Value, b.Value)
		})
		for _, bucket := range dimension.Values {
			if bucket.Count >= 10 && bucket.Count*5 >= census.collection.FileCount*4 {
				result.Spikes = append(result.Spikes, QualitySpike{field, bucket.Value, bucket.Count})
			}
		}
		if len(dimension.Values) > 50 {
			for _, bucket := range dimension.Values[50:] {
				dimension.Other += bucket.Count
			}
			dimension.Values = dimension.Values[:50]
		}
		result.Dimensions = append(result.Dimensions, dimension)
	}
	return result, nil
}

func qualityValue(row collectionQualityRow, field string, configured bool) string {
	switch field {
	case "extension":
		return strings.ToLower(strings.TrimPrefix(path.Ext(row.Name), "."))
	case "media_type":
		mediaType, _, err := mime.ParseMediaType(row.MediaType)
		if err != nil {
			return ""
		}
		return strings.ToLower(mediaType)
	case "media_family":
		family := query.ClassifyMedia(row.MediaType, row.Name)
		if family == "unknown" {
			return ""
		}
		return family
	case "modified_month":
		modified, err := time.Parse(time.RFC3339Nano, row.ModifiedAt)
		if err != nil {
			return ""
		}
		return modified.UTC().Format("2006-01")
	case "size":
		return collectionQualitySize(row.Size)
	case "text_coverage":
		if !configured {
			return ""
		}
		return row.State
	case "duplicates":
		if row.DuplicateCount > 1 {
			return "duplicate"
		}
		return "unique"
	default:
		return ""
	}
}

func collectionQualitySize(size int64) string {
	switch {
	case size == 0:
		return "zero"
	case size < 1<<10:
		return "(0,1KiB)"
	case size < 1<<20:
		return "[1KiB,1MiB)"
	case size < 10<<20:
		return "[1MiB,10MiB)"
	case size < 100<<20:
		return "[10MiB,100MiB)"
	case size < 1<<30:
		return "[100MiB,1GiB)"
	default:
		return ">=1GiB"
	}
}
